src/core/or/connection_or.c

```diff
@@ -1364,10 +1366,34 @@ void
 connection_or_connect_failed(or_connection_t *conn,
                              int reason, const char *msg)
 {
+  const or_options_t *options = get_options();
+
   control_event_or_conn_status(conn, OR_CONN_EVENT_FAILED, reason);
-  if (!authdir_mode_tests_reachability(get_options()))
+  if (!authdir_mode_tests_reachability(options))
     control_event_bootstrap_prob_or(msg, reason, conn);
   note_or_connect_failed(conn);
+
+  if (options->ClientAutoIPv6ORPort) {
+    or_state_t *state = get_or_state();
+    switch (tor_addr_family(&conn->real_addr)) {
+      case AF_INET:
+        if (reason == END_OR_CONN_REASON_NO_ROUTE) {
+          state->IPv4AutoFail = 1;
+          if (!state->IPv4AutoFailTime)
+            state->IPv4AutoFailTime = time(NULL);
+        }
+        state->IPv4Fails++;
+        break;
+      case AF_INET6:
+        if (reason == END_OR_CONN_REASON_NO_ROUTE) {
+          state->IPv6AutoFail = 1;
+          if (!state->IPv6AutoFailTime)
+            state->IPv6AutoFailTime = time(NULL);
+        }
+        state->IPv6Fails++;
```

Comment:
Incrementing an int can overflow. Instead, use uint32_t and tor_add_u32_nowrap().